LIVE-1443 LL-1170 Improve Bitcoin performance

hzheng-ledger opened the pull request:
1. use react-native-secp256k1 lib to calculate elliptic curve in async mode
2. get rid of unnecessary lib for zcash
3. add index to improve getLastTx

hzheng-ledger committed:
secp256k1 lib update

hzheng-ledger committed:
fix secp256k1 for react-native

hzheng-ledger committed:
improve secp256k1

hzheng-ledger committed:
refactoring secp256

hzheng-ledger committed:
secp256k1 lib update

hzheng-ledger committed:
secp256k1 lib update

haammar-ledger commented on src/families/bitcoin/wallet-btc/crypto/bip32.ts:
```diff
@@ -28,7 +29,13 @@ class BIP32 {
     const IR = I.slice(32);
-    const Ki = Buffer.from(publicKeyTweakAdd(this.publicKey, IL));
+    const Ki = Buffer.from(
+      typeof navigator !== "undefined" && navigator.product === "ReactNative"
```
Just wondering, would it not be good to have the new code on all platforms? Desktop and mobile?

hzheng-ledger replied:
no, the lib "secp256k1" is faster in LLD because of C++ but it is very slow in LLM as it takes js implementation as fallback plan：
so the lib "noble-secp256k1"(the new one added by me) is used by LLM only

gre replied:
the way we "feature detect" ReactNative here is not really safe to me. it could be better to introduce a "implementation" concept and a "setImplementation" that the client side will have to call.

gre replied:
this idea of having a separate setImplementation could come when we will really need it for when we inject `react-native-secp256k1` – meanwhile it's acceptable to take this quick win with the embedded branching – it could be good to add a comment on the line of code with:

```
// FIXME LIVE-1443
```

( see https://ledgerhq.atlassian.net/browse/LIVE-1443 )

haammar-ledger commented on src/families/bitcoin/wallet-btc/crypto/secp256k1.ts:
```diff
@@ -0,0 +1,420 @@
+/* eslint-disable */
+const BigInt = require('big-integer');
```
`big-integer` is currently available only through other dependencies, maybe we should add it to package.json?

Also I think we often had issues with bigint polyfills in the past... can you please make sure to test on both Android and iOS?

hzheng-ledger committed:
update secp256k1

hzheng-ledger committed:
update secp256k1

hzheng-ledger committed:
secp256k1 update

hzheng-ledger committed:
code refactoring

hzheng-ledger committed:
add account index to improve perf

hzheng-ledger committed:
code refactoring

hzheng-ledger committed:
remove unnecessary zec js lib

hzheng-ledger committed:
merge from develop

hzheng-ledger committed:
change secp256k1 into async mode

haammar-ledger commented on src/families/bitcoin/wallet-btc/crypto/zec.ts:
```diff
@@ -74,7 +71,20 @@ class ZCash extends Base {
-    return zec.Address.isValid(address, "livenet");
+    try {
+      const version = Number(
+        "0x" + bs58check.decode(address).slice(0, 2).toString("hex")
```
You can add a comment to document this

hzheng-ledger replied:
fixed

haammar-ledger replied:
where? :smile:

gre replied:
why is there something specific to zcash on the "performance" topic? how should QA test this? thanks

gre replied:
why is there something specific to zcash on the "performance" topic? how should QA test this? thanks

haammar-ledger reviewed (commented):
Please update yarn.lock

haammar-ledger commented on src/families/bitcoin/wallet-btc/storage/index.ts:
```diff
@@ -183,14 +206,23 @@ class BitcoinLikeStorage implements IStorage {
     this.addressCache = data.addressCache;
     Base.addressCache = { ...Base.addressCache, ...this.addressCache };
+    if (
+      (!this.accountIndex || _.isEmpty(this.accountIndex)) &&
```
Can you please try import isEmpty instead of _ from lodash ?

hzheng-ledger replied:
fixed

haammar-ledger commented:
Looks good :+1: but needs to test manually all BTC family on iOS + Android.

hzheng-ledger committed:
add some comment

hzheng-ledger committed:
fix conflict

hzheng-ledger committed:
use @ledgerhq/react-native-secp256k1 instead of react-native-secp256k1

hzheng-ledger committed:
fix conflict

gre commented on package.json:
```diff
@@ -68,6 +68,7 @@
     "@ledgerhq/json-bignumber": "^1.1.0",
     "@ledgerhq/live-app-sdk": "^0.2.0",
     "@ledgerhq/logs": "6.10.0",
+    "@ledgerhq/react-native-secp256k1": "^1.0.6",
```
here is a basic idea of what i mean (quick draft):

LLM

```js
import { setCryptoImpl } from "@ledgerhq/live-common/lib/some/path/crypto-impl";
import Something from "@ledgerhq/react-native-secp256k1";

setCryptoImpl({
  // well i don't know the contract of the crypto function you need, but basically give this impl an object OR just a function that would use Something. obviously it's async
})
```

LL Common

```js
// file src/some/path/crypto-impl.js

let impl = ...; here there need to be a "default" implementation, that JS would use basically. so our impl technically works in web for instance

// this allows to override default impl. LLM can inject its own impl, but LLD could have its own too (node.js one for instance)
export function setCryptoImpl (other) {
  impl = other;
}

export function yourCryptoFunction (...) {
// just use impl to "proxy" whatver you need
}
```

hzheng-ledger replied:
fixed

hzheng-ledger committed:
remove react-native-secp256k1

hzheng-ledger committed:
update secp256k1

hzheng-ledger committed:
fix batch size

hzheng-ledger committed:
small refactoring

hzheng-ledger commented on src/families/bitcoin/wallet-btc/crypto/secp256k1.ts:
```diff
@@ -0,0 +1,13 @@
+export async function publicKeyTweakAdd(
+  publicKey: Uint8Array,
+  tweak: Uint8Array
+): Promise<any> {
```
I've updated to Promise<number[]>. thanks

hzheng-ledger committed:
fix conflict

gre committed:
Merge branch 'develop' into feat/bitcoin-js-LL-1170

gre commented on src/families/bitcoin/wallet-btc/storage/index.ts:
```diff
@@ -20,23 +23,35 @@ class BitcoinLikeStorage implements IStorage {
-        (!has(txFilter, "address") || t.address === txFilter.address) &&
-        (!has(txFilter, "confirmed") ||
+    const tx: TX | undefined = findLast(
+      this.accountIndex[`${txFilter.account}-${txFilter.index}`].map(
```
please avoid all the repetition maybe with a `hashAccountIndex(tx)` funciton shared across every parts.

gre replied:
please avoid all the repetition maybe with a `hashAccountIndex(tx)` funciton shared across every parts.

gre committed:
Fixes for the async implementation of wallet-btc (#1821)

- improve typing of secp256k1. contract changed: setSecp256k1Instance + getSecp256k1Instance
- This rework also fixes a bug we had on mobile. LIVE-1716 but we will have to change the function call to use setSecp256k1Instance
- improve logic of getPubkeyAt caches. the promises are cached and not their result. this ensure concurrency of calls will wait for the same calculation and avoid some congestion of calls. documented it a bit more.
- add some invariant() to prevent future case of empty publicKey

gre committed:
add test that provde error propagate correctly
